File: execviz-erl/execviz.erl
% =========================================================================
%  MANIFEST
% =========================================================================
%  script_name: execviz.erl
%  script_path: execviz-erl/execviz.erl
%  module_name: execviz
%  version: 0.53.1
%  description: %% execviz capture adapter for the BEAM.
%  kind: module
%  spec: internal
%  internal_dependencies: 
%  external_dependencies: 
%  features: execviz, capture, adapter
%  api_version: execvis-v1.0.0
%  last_updated: 2026-08-07
% =========================================================================

%%% execviz capture adapter for the BEAM.
%%%
%%% The BEAM's carrier problem is the cleanest of any runtime supported so far,
%%% and the least forgiving. A process has its own heap and its own process
%%% dictionary, so the dictionary IS a correct per-process carrier: nothing
%%% leaks between concurrent processes because nothing is shared.
%%%
%%% What it does not do is cross a `spawn` or a message send. A spawned process
%%% starts with an empty dictionary, which is right; it is a different unit of
%%% execution; so the parent span is carried across explicitly, exactly as a
%%% thread boundary is in Ruby and a fiber boundary is in PHP. Inheriting it
%%% silently would be the same mistake as using a frame stack across an await.
-module(execviz).

-export([install/2, install/3, set_domain/1, current/0,
         span_start/2, span_start/3, span_end/2, span_end/3,
         with_span/3, spawn_span/2, gather/2,
         stamp/1, claim/1, release/1, span_event/3,
         lifecycle/2, log/2, flush/0]).

-define(STATE, execviz_state).
-define(SPAN, execviz_span).
-define(BATCH, 200).

%% ========================================================================
%% LIFECYCLE
%% ========================================================================

install(Collector, HostId) -> install(Collector, HostId, <<"app">>).

install(Collector, HostId, Domain) ->
    Pid = case whereis(execviz_buffer) of
        undefined ->
            P = spawn(fun() -> buffer_loop(Collector, HostId, [], #{}) end),
            register(execviz_buffer, P),
            P;
        Existing -> Existing
    end,
    put(?STATE, #{collector => Collector, host => HostId, domain => Domain,
                  trace => sid(), buffer => Pid}),
    Pid.

set_domain(D) ->
    S = state(), put(?STATE, S#{domain => D}), ok.

state() ->
    case get(?STATE) of
        undefined ->
            %% a process that was spawned without carrying context still records
            %% accurately rather than crashing: it inherits nothing and reports it
            #{collector => "http://127.0.0.1:8900", host => <<"beam">>,
              domain => <<"unknown">>, trace => sid(), buffer => whereis(execviz_buffer)};
        S -> S
    end.

current() -> case get(?SPAN) of undefined -> null; S -> S end.

%% ========================================================================
%% SPANS
%% ========================================================================

span_start(Name, Kind) -> span_start(Name, Kind, []).

span_start(Name, Kind, Opts) ->
    S = state(),
    Id = sid(),
    Parent = proplists:get_value(parent, Opts, current()),
    Links = proplists:get_value(links, Opts, []),
    Domain = proplists:get_value(domain, Opts, maps:get(domain, S)),
    Span = #{span_id => Id, trace_id => maps:get(trace, S),
             parent_span_id => Parent, links => Links,
             name => to_bin(Name), kind => to_bin(Kind),
             start => now_secs(), 'end' => null, status => <<"running">>,
             lifecycle => [], events => [], origin => <<"semantic">>,
             host_id => maps:get(host, S), domain => Domain,
             attributes => #{pid => to_bin(pid_to_list(self()))}},
    send({span, Span}),
    Id.

span_end(Id, Status) -> span_end(Id, Status, #{}).

span_end(Id, Status, Attrs) ->
    send({finish, Id, now_secs(), to_bin(Status), Attrs}),
    ok.

lifecycle(Id, Type) -> send({lifecycle, Id, now_secs(), to_bin(Type)}), ok.

%% A log line belongs to the span that was running when it was written.
%% Attaches a line to a named span, whatever is current.
%%
%% Exists for the logger handler, which runs in the caller's process
%% and has already resolved the span before calling: passing it explicitly means
%% the handler never has to assume the dictionary is still what it was.
span_event(Span, Level, Msg) ->
    send({event, Span, now_secs(), to_bin(Level), to_bin(Msg)}), ok.

log(Level, Msg) ->
    case current() of
        null -> ok;
        Id -> send({event, Id, now_secs(), to_bin(Level), to_bin(Msg)}), ok
    end.

%% Runs Fun inside a span, with that span active for whatever it reaches.
with_span(Name, Kind, Fun) ->
    Id = span_start(Name, Kind),
    Prev = get(?SPAN),
    put(?SPAN, Id),
    try
        R = Fun(),
        span_end(Id, ok),
        R
    catch
        Class:Reason:Stack ->
            span_end(Id, error, #{error => to_bin(io_lib:format("~p:~p", [Class, Reason])),
                                  stack => to_bin(io_lib:format("~p", [lists:sublist(Stack, 5)]))}),
            erlang:raise(Class, Reason, Stack)
    after
        case Prev of undefined -> erase(?SPAN); _ -> put(?SPAN, Prev) end
    end.

%% A spawn is a crossing. The child inherits the parent span because it is
%% handed it here, not because the runtime shares anything: a spawned process
%% starts with an empty dictionary, and that is correct.
spawn_span(Name, Fun) ->
    S = state(),
    Parent = current(),
    Id = span_start(Name, spawn),
    Pid = spawn(fun() ->
        put(?STATE, S),
        put(?SPAN, Id),
        try Fun(), span_end(Id, ok)
        catch C:R:_ -> span_end(Id, error, #{error => to_bin(io_lib:format("~p:~p",[C,R]))})
        end
    end),
    _ = Parent,
    {Pid, Id}.

%% A fan-in: the join keeps the enclosing scope as its parent and names every
%% child in links. Parenting it to a child would place it outside
%% its own parent in time.
gather(Name, Funs) ->
    Parent = current(),
    Self = self(),
    Ids = [begin
        Id = span_start(io_lib:format("~s[~p]", [Name, I]), call, [{parent, Parent}]),
        spawn(fun() ->
            put(?SPAN, Id),
            R = try Fun(), ok catch C:Rn:_ -> {error, {C, Rn}} end,
            span_end(Id, case R of ok -> ok; _ -> error end),
            Self ! {done, Id}
        end),
        Id
    end || {I, Fun} <- lists:zip(lists:seq(0, length(Funs) - 1), Funs)],
    [receive {done, _} -> ok after 30000 -> timeout end || _ <- Ids],
    Join = span_start(io_lib:format("~s_join", [Name]), call, [{parent, Parent}, {links, Ids}]),
    span_end(Join, ok),
    Ids.

%% ========================================================================
%% CROSSINGS
%% ========================================================================

%% Context stamped onto a message, read back on the far side. A message send is
%% the BEAM's boundary, and it is explicit here because it is explicit there.
stamp(Msg) ->
    S = state(),
    {execviz, maps:get(trace, S), current(), Msg}.

claim({execviz, Trace, Span, Msg}) ->
    S = state(),
    put(?STATE, S#{trace => Trace}),
    case Span of
        null -> ok;
        _ -> lifecycle(Span, claimed), put(?SPAN, Span)
    end,
    {Msg, Span};
claim(Other) -> {Other, null}.

release(null) -> ok;
release(Span) -> lifecycle(Span, released), span_end(Span, ok), erase(?SPAN), ok.

%% ========================================================================
%% DELIVERY
%% ========================================================================

send(Msg) ->
    case whereis(execviz_buffer) of
        undefined -> ok;          %% never crash the program being observed
        P -> P ! Msg, ok
    end.

flush() ->
    case whereis(execviz_buffer) of
        undefined -> ok;
        P -> P ! {flush, self()}, receive {flushed, N} -> N after 10000 -> timeout end
    end.

%% The buffer is one process holding every span, which is the BEAM way of having
%% shared state without a lock. Two-phase writes are applied here, so the far end
%% receives the same span twice and upserts it.
buffer_loop(Collector, Host, Pending, Sent) ->
    receive
        {span, S} ->
            buffer_loop(Collector, Host, [S | Pending], Sent);
        {finish, Id, T, Status, Attrs} ->
            P2 = [case maps:get(span_id, S) of
                      Id -> S#{'end' => T, status => Status,
                               attributes => maps:merge(maps:get(attributes, S), Attrs)};
                      _ -> S
                  end || S <- Pending],
            buffer_loop(Collector, Host, P2, Sent);
        {lifecycle, Id, T, Type} ->
            P2 = [case maps:get(span_id, S) of
                      Id -> S#{lifecycle => maps:get(lifecycle, S) ++ [#{t => T, type => Type}]};
                      _ -> S
                  end || S <- Pending],
            buffer_loop(Collector, Host, P2, Sent);
        {event, Id, T, Level, Msg} ->
            P2 = [case maps:get(span_id, S) of
                      Id -> S#{events => maps:get(events, S) ++ [#{t => T, level => Level, msg => Msg}]};
                      _ -> S
                  end || S <- Pending],
            buffer_loop(Collector, Host, P2, Sent);
        {flush, From} ->
            {N, Sent2} = deliver(Collector, Host, Pending, Sent),
            From ! {flushed, N},
            buffer_loop(Collector, Host, Pending, Sent2)
    after 1000 ->
        {_, Sent2} = deliver(Collector, Host, Pending, Sent),
        buffer_loop(Collector, Host, Pending, Sent2)
    end.

%% A span is re-sent once its second phase lands, and not otherwise.
deliver(_Collector, _Host, [], Sent) -> {0, Sent};
deliver(Collector, Host, Pending, Sent) ->
    Changed = [S || S <- Pending,
        maps:get(phase_key(S), Sent, undefined) =/= phase_val(S)],
    case Changed of
        [] -> {0, Sent};
        _ ->
            Body = json_obj([{host_id, Host}, {spans, [span_json(S) || S <- Changed]}]),
            case post(Collector ++ "/api/ingest", Body) of
                ok ->
                    S2 = lists:foldl(fun(S, Acc) ->
                        maps:put(phase_key(S), phase_val(S), Acc) end, Sent, Changed),
                    {length(Changed), S2};
                _ -> {0, Sent}      %% a failed delivery is retried, never dropped
            end
    end.

phase_key(S) -> maps:get(span_id, S).
phase_val(S) -> {maps:get('end', S), maps:get(status, S)}.

%% ========================================================================
%% PLUMBING
%% ========================================================================

post(Url, Body) ->
    case parse_url(Url) of
        {Host, Port, Path} ->
            case gen_tcp:connect(Host, Port, [binary, {active, false}, {packet, 0}], 5000) of
                {ok, Sock} ->
                    Req = iolist_to_binary([
                        "POST ", Path, " HTTP/1.1\r\nHost: ", Host, "\r\n",
                        "Content-Type: application/json\r\n",
                        "Content-Length: ", integer_to_list(byte_size(Body)), "\r\n",
                        "Connection: close\r\n\r\n", Body]),
                    gen_tcp:send(Sock, Req),
                    Reply = case gen_tcp:recv(Sock, 0, 5000) of
                                {ok, R} -> R;
                                _ -> <<>>
                            end,
                    gen_tcp:close(Sock),
                    report_refusals(Reply),
                    ok;
                _ -> error
            end;
        _ -> error
    end.

%% Reads what the collector said about the batch.
%%
%% It names every span it refused and why. That explanation reached nobody while
%% the reply was discarded and any answer treated as complete success, so an
%% adapter emitting malformed spans went on emitting them with nothing to show
%% its author.
%%
%% Reported once per distinct reason: a bug in an adapter repeats every second,
%% and a message that repeats with it is one nobody reads.
report_refusals(Reply) when is_binary(Reply) ->
    %% Whitespace-tolerant on purpose: assuming a peer formats JSON compactly is
    %% an assumption about someone else's serialiser, and it fails silently,
    %% the reply is read, nothing matches, and no refusal is ever reported.
    case rejected_count(Reply) of
        0 -> ok;
        _ ->
            case binary:match(Reply, <<"\"reasons\"">>) of
                nomatch -> ok;
                {S, _} ->
                    Tail = binary:part(Reply, S, byte_size(Reply) - S),
                    case {binary:match(Tail, <<"[">>), binary:match(Tail, <<"]">>)} of
                        {{O, _}, {C, _}} when C > O ->
                            Inner = binary:part(Tail, O + 1, C - O - 1),
                            lists:foreach(fun report_one/1,
                                          binary:split(Inner, <<"\",">>, [global]));
                        _ -> ok
                    end
            end
    end;
report_refusals(_) -> ok.

%% The count, whatever spacing the peer used after the colon.
rejected_count(Reply) ->
    case binary:match(Reply, <<"\"rejected\"">>) of
        nomatch -> 0;
        {S, L} ->
            Tail = binary:part(Reply, S + L, min(24, byte_size(Reply) - S - L)),
            Digits = [C || <<C>> <= Tail, C >= $0, C =< $9],
            case Digits of
                [] -> 0;
                _ -> list_to_integer(Digits)
            end
    end.

report_one(Raw) ->
    Reason = binary:replace(Raw, <<"\"">>, <<>>, [global]),
    case Reason of
        <<>> -> ok;
        _ ->
            %% the span id changes every time, so key on the explanation itself
            Key = case binary:split(Reason, <<":">>) of
                      [_, Rest] -> Rest;
                      _ -> Reason
                  end,
            Seen = case get(execviz_reported) of undefined -> []; L -> L end,
            case lists:member(Key, Seen) of
                true -> ok;
                false ->
                    put(execviz_reported, [Key | Seen]),
                    io:format(standard_error,
                              "execviz: the collector refused a span; ~s~n"
                              "  (further spans refused for this reason will not be reported again)~n",
                              [Reason])
            end
    end.

parse_url("http://" ++ Rest) ->
    {HostPort, Path} = case string:split(Rest, "/") of
        [HP] -> {HP, "/"};
        [HP, Tail] -> {HP, "/" ++ Tail}
    end,
    case string:split(HostPort, ":") of
        [H, PortStr] -> {H, list_to_integer(PortStr), Path};
        [H] -> {H, 80, Path}
    end;
parse_url(_) -> error.

span_json(S) ->
    json_obj([{span_id, maps:get(span_id, S)}, {trace_id, maps:get(trace_id, S)},
              {parent_span_id, maps:get(parent_span_id, S)},
              {links, maps:get(links, S)}, {name, maps:get(name, S)},
              {kind, maps:get(kind, S)}, {start, maps:get(start, S)},
              {'end', maps:get('end', S)}, {status, maps:get(status, S)},
              {lifecycle, [json_obj([{t, maps:get(t,L)},{type, maps:get(type,L)}]) || L <- maps:get(lifecycle, S)]},
              {events, [json_obj([{t, maps:get(t,E)},{level, maps:get(level,E)},{msg, maps:get(msg,E)}]) || E <- maps:get(events, S)]},
              {origin, maps:get(origin, S)},
              %% which clock stamped this, so skew analysis knows what it compares
              {clock_source, <<"erlang:system_time">>},
              {host_id, maps:get(host_id, S)},
              {domain, maps:get(domain, S)},
              {attributes, json_obj(maps:to_list(maps:get(attributes, S)))}]).

json_obj(KVs) ->
    iolist_to_binary(["{", lists:join(",", [[jkey(K), ":", jval(V)] || {K, V} <- KVs]), "}"]).

jkey(K) -> ["\"", esc(to_bin(K)), "\""].

jval(null) -> "null";
jval(true) -> "true";
jval(false) -> "false";
jval(V) when is_integer(V) -> integer_to_list(V);
jval(V) when is_float(V) -> io_lib:format("~.6f", [V]);
jval(V) when is_binary(V) ->
    case V of
        <<"{", _/binary>> -> V;          %% already-rendered JSON
        _ -> ["\"", esc(V), "\""]
    end;
jval(V) when is_list(V) ->
    case io_lib:printable_list(V) of
        true -> ["\"", esc(to_bin(V)), "\""];
        false -> ["[", lists:join(",", [jval(X) || X <- V]), "]"]
    end;
jval(V) when is_atom(V) -> ["\"", esc(to_bin(V)), "\""];
jval(V) -> ["\"", esc(to_bin(io_lib:format("~p", [V]))), "\""].

esc(B) when is_binary(B) ->
    << <<(case C of
              $" -> <<"\\\"">>;
              $\\ -> <<"\\\\">>;
              $\n -> <<"\\n">>;
              $\r -> <<"\\r">>;
              $\t -> <<"\\t">>;
              _ -> <<C>>
          end)/binary>> || <<C>> <= B >>.

to_bin(B) when is_binary(B) -> B;
to_bin(A) when is_atom(A) -> atom_to_binary(A, utf8);
to_bin(I) when is_integer(I) -> integer_to_binary(I);
to_bin(L) when is_list(L) -> iolist_to_binary(L).

now_secs() -> erlang:system_time(microsecond) / 1000000.

sid() ->
    <<A:48>> = crypto:strong_rand_bytes(6),
    list_to_binary(io_lib:format("~12.16.0b", [A])).

File: execviz-dotnet/Execviz.cs
// ========================================================================
//  MANIFEST
// ========================================================================
//  script_name: Execviz.cs
//  script_path: execviz-dotnet/Execviz.cs
//  module_name: Execviz
//  version: 0.53.1
//  description: execviz capture adapter for .NET.
//  kind: module
//  spec: internal
//  internal_dependencies: 
//  external_dependencies: 
//  features: Execviz, capture, adapter
//  api_version: execvis-v1.0.0
//  last_updated: 2026-08-07
// ========================================================================

// execviz capture adapter for .NET.
//
// The carrier is AsyncLocal<T>. Like Python's contextvars and Node's
// AsyncLocalStorage it flows with the logical call, not with the thread, which
// makes it correct across `await`: a continuation resumed on a pooled
// thread still sees the span its caller started.
//
// The failure mode worth naming is the opposite of the JVM's. A ThreadLocal on
// a pooled thread *leaks* a stale span into unrelated work; AsyncLocal does not,
// because the value is captured into the ExecutionContext at the await point
// rather than living on the thread. What AsyncLocal does not do is flow *out*
// of a context; a value set inside an async method is invisible to its caller,
// because every span is opened and closed within one scope here.
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Diagnostics;
using System.Globalization;
using System.Linq;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.IO;
using System.Threading;
using System.Threading.Tasks;

namespace Execviz
{
    public sealed class Span
    {
        public string SpanId { get; set; } = "";
        public string TraceId { get; set; } = "";
        public string? ParentSpanId { get; set; }
        public List<string> Links { get; } = new();
        public string Name { get; set; } = "";
        public string Kind { get; set; } = "call";
        public double Start { get; set; }
        public double? End { get; set; }
        public string Status { get; set; } = "running";
        public string HostId { get; set; } = "dotnet";
        public string? Domain { get; set; }
        public List<object> Lifecycle { get; } = new();
        public List<object> Events { get; } = new();
        public Dictionary<string, object?> Attributes { get; } = new();
        public object? Error { get; set; }
    }

    public static class Capture
    {
        // AsyncLocal flows with the logical call. This is the whole adapter's
        // correctness resting on one type.
        private static readonly AsyncLocal<string?> Current = new();
        private static readonly AsyncLocal<string?> Trace = new();

        private static readonly ConcurrentDictionary<string, Span> Pending = new();
        private static readonly ConcurrentDictionary<string, string> Sent = new();
        private static readonly HttpClient Http = new() { Timeout = TimeSpan.FromSeconds(8) };
        private static string _collector = "http://127.0.0.1:8900";
        private static string _host = "dotnet-1";
        private static string _domain = "app";
        private static Timer? _timer;

        public static void Install(string? collector = null, string host = "dotnet-1", string domain = "app")
        {
            _collector = collector
                ?? Environment.GetEnvironmentVariable("EXECVIZ_COLLECTOR")
                ?? "http://127.0.0.1:8900";
            _host = host;
            _domain = domain;
            Trace.Value = Sid();
            // periodic delivery, and an explicit Flush for shutdown: a process
            // that exits without flushing loses exactly the record of why
            _timer = new Timer(_ => Deliver(), null, 700, 700);
        }

        public static void SetDomain(string d) => _domain = d;
        public static string? CurrentSpan() => Current.Value;

        public static string SpanStart(string name, string kind = "call",
            string? parent = null, IEnumerable<string>? links = null,
            string? domain = null, IDictionary<string, object?>? attributes = null)
        {
            var id = Sid();
            var s = new Span
            {
                SpanId = id,
                TraceId = Trace.Value ??= Sid(),
                ParentSpanId = parent ?? Current.Value,
                Name = name,
                Kind = kind,
                Start = Now(),
                HostId = _host,
                Domain = domain ?? _domain,
            };
            if (links != null) s.Links.AddRange(links);
            s.Attributes["thread"] = Environment.CurrentManagedThreadId;
            if (attributes != null) foreach (var kv in attributes) s.Attributes[kv.Key] = kv.Value;
            Pending[id] = s;
            return id;
        }

        public static void SpanEnd(string id, string status = "ok",
            IDictionary<string, object?>? attributes = null, Exception? error = null)
        {
            if (!Pending.TryGetValue(id, out var s)) return;
            s.End = Now();
            s.Status = status;
            if (attributes != null) foreach (var kv in attributes) s.Attributes[kv.Key] = kv.Value;
            if (error != null) s.Error = DescribeError(error);
        }

        public static void Lifecycle(string id, string type, object? context = null)
        {
            if (!Pending.TryGetValue(id, out var s)) return;
            var e = new Dictionary<string, object?> { ["t"] = Now(), ["type"] = type };
            if (context != null) e["context"] = context;
            s.Lifecycle.Add(e);
        }

        /// A log line belongs to the span that was running when it was written.
        public static void Log(string level, string message)
        {
            var id = Current.Value;
            if (id == null || !Pending.TryGetValue(id, out var s)) return;
            s.Events.Add(new Dictionary<string, object?>
                { ["t"] = Now(), ["level"] = level, ["msg"] = message });
        }

        /// Type, message, frames and the cause chain beneath.
        /// The chain matters most: the exception on top is usually the least
        /// informative one in the stack.
        private static object DescribeError(Exception e)
        {
            var frames = new StackTrace(e, true).GetFrames()?
                .Take(12)
                .Select(f => new Dictionary<string, object?>
                {
                    ["file"] = f.GetFileName(),
                    ["line"] = f.GetFileLineNumber(),
                    ["func"] = f.GetMethod()?.Name,
                })
                .ToList() ?? new List<Dictionary<string, object?>>();
            var chain = new List<object>();
            var cur = e.InnerException;
            var guard = 0;
            while (cur != null && guard++ < 8)
            {
                chain.Add(new Dictionary<string, object?>
                    { ["type"] = cur.GetType().Name, ["message"] = cur.Message });
                cur = cur.InnerException;
            }
            var err = new Dictionary<string, object?>
            {
                ["type"] = e.GetType().Name,
                ["message"] = e.Message,
                ["frames"] = frames,
            };
            if (chain.Count > 0) err["caused_by"] = chain;
            return err;
        }

        /// Runs the work inside a span, with that span active for what it reaches.
        /// The span is opened and closed in one scope because AsyncLocal does not
        /// flow out of an async method back to its caller.
        public static async Task<T> WithSpan<T>(string name, string kind, Func<Task<T>> work,
            string? domain = null)
        {
            var id = SpanStart(name, kind, domain: domain);
            var prev = Current.Value;
            Current.Value = id;
            try
            {
                var r = await work().ConfigureAwait(false);
                SpanEnd(id, "ok");
                return r;
            }
            catch (Exception ex)
            {
                SpanEnd(id, "error", error: ex);
                throw;
            }
            finally { Current.Value = prev; }
        }

        /// Makes an already-open span current for the duration of some work.
        /// AsyncLocal does not flow out of an async method back to its caller,
        /// so a span opened on behalf of a caller has to be made current around
        /// the work rather than merely assigned.
        public static async Task WithSpanCurrent(string spanId, Func<Task> work)
        {
            var prev = Current.Value;
            Current.Value = spanId;
            try { await work().ConfigureAwait(false); }
            finally { Current.Value = prev; }
        }

        public static Task WithSpan(string name, string kind, Func<Task> work, string? domain = null)
            => WithSpan<object?>(name, kind, async () => { await work().ConfigureAwait(false); return null; }, domain);

        /// A fan-in: the join keeps the enclosing scope as its parent and names
        /// every child in links. Parenting it to a child would place
        /// it outside its own parent in time.
        public static async Task Gather(string name, params Func<Task>[] work)
        {
            var parent = Current.Value;
            var ids = new List<string>();
            var tasks = new List<Task>();
            for (var i = 0; i < work.Length; i++)
            {
                var id = SpanStart($"{name}[{i}]", "call", parent: parent);
                ids.Add(id);
                var w = work[i];
                tasks.Add(Task.Run(async () =>
                {
                    Current.Value = id;
                    try { await w().ConfigureAwait(false); SpanEnd(id, "ok"); }
                    catch (Exception ex) { SpanEnd(id, "error", error: ex); }
                }));
            }
            await Task.WhenAll(tasks).ConfigureAwait(false);
            var join = SpanStart($"{name}_join", "call", parent: parent, links: ids);
            SpanEnd(join, "ok");
        }

        /// Context stamped onto a value crossing a boundary, read back on the
        /// far side. A queue is where a carrier cannot follow, so it is explicit.
        public static Dictionary<string, object?> Stamp(object? item) => new()
        {
            ["item"] = item, ["trace_id"] = Trace.Value, ["span"] = Current.Value,
        };

        public static (object? Item, string? QueueSpan) Claim(Dictionary<string, object?> msg)
        {
            var qs = msg.TryGetValue("span", out var v) ? v as string : null;
            if (msg.TryGetValue("trace_id", out var t) && t is string tid) Trace.Value = tid;
            if (qs != null)
            {
                Lifecycle(qs, "claimed");
                Current.Value = qs;
            }
            return (msg.TryGetValue("item", out var it) ? it : null, qs);
        }

        public static void Release(string? queueSpan)
        {
            if (queueSpan == null) return;
            Lifecycle(queueSpan, "released");
            SpanEnd(queueSpan, "ok");
            Current.Value = null;
        }

        /// A span is re-sent once its second phase lands; a failed delivery is
        /// retried rather than dropped.
        public static void Deliver()
        {
            var batch = Pending.Values
                .Where(s => !Sent.TryGetValue(s.SpanId, out var st) || st != PhaseOf(s))
                .ToList();
            if (batch.Count == 0) return;
            try
            {
                var body = JsonSerializer.Serialize(new Dictionary<string, object?>
                {
                    ["host_id"] = _host,
                    ["spans"] = batch.Select(ToWire).ToList(),
                });
                var res = Http.PostAsync(_collector + "/api/ingest",
                    new StringContent(body, Encoding.UTF8, "application/json")).GetAwaiter().GetResult();
                if (!res.IsSuccessStatusCode) return;
                ReportRefusals(res.Content.ReadAsStringAsync().GetAwaiter().GetResult());
                foreach (var s in batch)
                {
                    Sent[s.SpanId] = PhaseOf(s);
                    if (s.End != null) Pending.TryRemove(s.SpanId, out _);
                }
            }
            catch { /* never fail the program being observed because recording failed */ }
        }

        /// <summary>
        /// Reads what the collector said about the batch.
        ///
        /// It names every span it refused and why. That explanation reached
        /// nobody while the reply was discarded and any 200 treated as complete
        /// success, so an adapter emitting malformed spans went on emitting them
        /// with nothing to show its author.
        ///
        /// Reported once per distinct reason: a bug in an adapter repeats every
        /// second, and a message that repeats with it is one nobody reads.
        /// </summary>
        private static readonly ConcurrentDictionary<string, bool> ReportedRefusals = new();
        private static long _refused;

        public static long RefusedByCollector() => Interlocked.Read(ref _refused);

        private static void ReportRefusals(string body)
        {
            try
            {
                using var doc = JsonDocument.Parse(body);
                if (!doc.RootElement.TryGetProperty("rejected", out var r) || r.GetInt32() <= 0) return;
                Interlocked.Add(ref _refused, r.GetInt32());
                if (!doc.RootElement.TryGetProperty("reasons", out var reasons)) return;
                foreach (var el in reasons.EnumerateArray())
                {
                    var reason = el.GetString();
                    if (string.IsNullOrEmpty(reason)) continue;
                    // the span id changes every time, so key on the explanation itself
                    var colon = reason.IndexOf(':');
                    var key = colon >= 0 ? reason[(colon + 1)..].Trim() : reason;
                    if (!ReportedRefusals.TryAdd(key, true)) continue;
                    Console.Error.WriteLine($"execviz: the collector refused a span; {reason}");
                    Console.Error.WriteLine("  (further spans refused for this reason will not be reported again)");
                }
            }
            catch
            {
                // an unreadable reply must never break delivery
            }
        }

        public static void Flush()
        {
            _timer?.Change(Timeout.Infinite, Timeout.Infinite);
            Deliver();
        }

        private static string PhaseOf(Span s) => (s.End != null ? "1" : "0") + "|" + s.Status;

        private static Dictionary<string, object?> ToWire(Span s)
        {
            var d = new Dictionary<string, object?>
            {
                ["span_id"] = s.SpanId,
                ["trace_id"] = s.TraceId,
                ["parent_span_id"] = s.ParentSpanId,
                ["links"] = s.Links,
                ["name"] = s.Name,
                ["kind"] = s.Kind,
                ["start"] = s.Start,
                ["end"] = s.End,
                ["status"] = s.Status,
                ["lifecycle"] = s.Lifecycle,
                ["events"] = s.Events,
                ["origin"] = "semantic",
                // which clock stamped this, so skew analysis knows what it compares
                ["clock_source"] = "DateTimeOffset.UtcNow",
                ["host_id"] = s.HostId,
                ["domain"] = s.Domain,
                ["attributes"] = s.Attributes,
            };
            if (s.Error != null) d["error"] = s.Error;
            return d;
        }

        private static double Now() =>
            DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() / 1000.0;

        private static string Sid()
        {
            Span<byte> b = stackalloc byte[6];
            System.Security.Cryptography.RandomNumberGenerator.Fill(b);
            return Convert.ToHexString(b).ToLower(CultureInfo.InvariantCulture);
        }
    }
}
